### ISSUE TITLE
ssl Handshake failure while connecting to devices when using python 3.10.2

### ISSUE THREAD

**Issue by shachiagarwal** [arista-eosplus/pyeapi#222]
Here is the failure log:
/usr/local/lib/python3.10/site-packages/vane-1.0.0-py3.10.egg/vane/tests_tools.py:817: in return_show_cmd
    show_output = conn.enable(show_cmd)
/usr/local/lib/python3.10/site-packages/pyeapi/client.py:712: in enable
    resp = self.run_commands(command, encoding, send_enable,
/usr/local/lib/python3.10/site-packages/pyeapi/client.py:771: in run_commands
    response = self._connection.execute(commands, encoding, **kwargs)
/usr/local/lib/python3.10/site-packages/pyeapi/eapilib.py:554: in execute
    response = self.send(request)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = EapiConnection(transport=[https://10.255.74.38:443//command-api](https://10.255.74.38//command-api))
data = b'{"jsonrpc": "2.0", "method": "runCmds", "params": {"version": 1, "cmds": ["enable", "show agent logs crash"], "format": "json"}, "id": "140108890278976", "streaming": false}'
    def send(self, data):
        """Sends the eAPI request to the destination node
    
        This method is responsible for sending an eAPI request to the
        destination node and returning a response based on the eAPI response
        object.  eAPI responds to request messages with either a success
        message or failure message.
    
        eAPI Response - success
    
        .. code-block:: json
    
            {
                "jsonrpc": "2.0",
                "result": [
                    {},
                    {}
                    {
                        "warnings": [
                            <message>
                        ]
                    },
                ],
                "id": <reqid>
            }
    
        eAPI Response - failure
    
        .. code-block:: json
    
            {
                "jsonrpc": "2.0",
                "error": {
                    "code": <int>,
                    "message": <string>
                    "data": [
                        {},
                        {},
                        {
                            "errors": [
                                <message>
                            ]
                        }
                    ]
                }
                "id": <reqid>
            }
    
        Args:
            data (string): The data to be included in the body of the eAPI
                request object
    
        Returns:
            A decoded response.  The response object is deserialized from
                JSON and returned as a standard Python dictionary object
    
        Raises:
            CommandError if an eAPI failure response object is returned from
                the node.   The CommandError exception includes the error
                code and error message from the eAPI response.
        """
        try:
            _LOGGER.debug('Request content: {}'.format(data))
            # debug('eapi_request: %s' % data)
    
            self.transport.putrequest('POST', '/command-api')
    
            self.transport.putheader('Content-type', 'application/json-rpc')
            self.transport.putheader('Content-length', '%d' % len(data))
    
            if self._auth:
                self.transport.putheader('Authorization',
                                         'Basic %s' % self._auth)
    
            if int(sys.version[0]) > 2:
                # For Python 3.x compatibility
                data = data.encode()
    
            self.transport.endheaders(message_body=data)
    
            try:  # Python 2.7: use buffering of HTTP responses
                response = self.transport.getresponse(buffering=True)
            except TypeError:  # Python 2.6: older, and 3.x on
                response = self.transport.getresponse()
    
            response_content = response.read()
            _LOGGER.debug('Response: status:{status}, reason:{reason}'.format(
                          status=response.status,
                          reason=response.reason))
            _LOGGER.debug('Response content: {}'.format(response_content))
    
            if response.status == 401:
                raise ConnectionError(str(self), '%s. %s' % (response.reason,
                                                             response_content))
    
            # Work around for Python 2.7/3.x compatibility
            if not type(response_content) == str:
                # For Python 3.x - decode bytes into string
                response_content = response_content.decode()
            decoded = json.loads(response_content)
            _LOGGER.debug('eapi_response: %s' % decoded)
    
            if 'error' in decoded:
                (code, msg, err, out) = self._parse_error_message(decoded)
                pattern = "unexpected keyword argument '(.*)'"
                match = re.search(pattern, msg)
                if match:
                    auto_msg = ('%s parameter is not supported in this'
                                ' version of EOS.' % match.group(1))
                    _LOGGER.error(auto_msg)
                    msg = msg + '. ' + auto_msg
                raise CommandError(code, msg, command_error=err, output=out)
    
            return decoded
    
        # socket.error is deprecated in python 3 and replaced with OSError.
        except (socket.error, OSError) as exc:
            _LOGGER.exception(exc)
            self.socket_error = exc
            self.error = exc
            error_msg = 'Socket error during eAPI connection: %s' % str(exc)
>           raise ConnectionError(str(self), error_msg)
E           pyeapi.eapilib.ConnectionError: Socket error during eAPI connection: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)
/usr/local/lib/python3.10/site-packages/pyeapi/eapilib.py:483: ConnectionError

Probably it is hitting this problem and we need to change pyeapi code? https://stackoverflow.com/questions/56719290/how-to-solve-the-handshake-failure-using-ssl-in-python

**Comment by dlyssenko** [comment]
The error `[SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)` in python 3.10x (and higher) is not a bug. It's a security strength improvement, which unfortunately, is not forward-compatible with the scripts developed under prior python versions (3.9 and below) which rely on https transport.

In a nutshell: python 3.10 increased default security settings of the TLS stack, which renders the https web servers that use older (weaker) key exchange fail (python would indicate that failure with the above error). There are 2 ways to deal with this issue:

a. configure https web server to use a stronger cipher suite (more secure key exchange) - however this might be beyond control of the python developer.
b. modify the python script allowing a weaker key exchange, namely for `pyeapi` it might look like this: 
```
import pyeapi
cc = pyeapi.client.connect( transport='https', host=host_name, username='admin')
cc.transport._context.set_ciphers('DEFAULT')
c = pyeapi.client.Node( cc )
c.enable('show version')
```

**Comment by gmuloc** [comment]
Thanks @dlyssenko  - I agree that this is not a bug,

I would argue the two following points:
1/ that `DEFAULT` is dangerous as you may authorise the usage of weak ciphers and so it SHOULD be recommended with caution.
2/ This would be great to have this documented in pyeapi documentation as the knowledge that you can access the underlying context using `._context` on the transport may not be know to a regular user.

**Comment by dlyssenko** [comment]
@gmuloc, The only correct remedy for this issue is `a.`. Whatever cipher you put in an option `b.` (`'DEFAULT'`, `'RSA'`, or anything else that would work) -  w/o addressing `a.` - is a hack and a security concern, because anyways it would lower the security strength required by python 3.10. Thus, the option `b.` must be considered only if option "a." is unattainable and with understanding the implications.
I'll update the pyeapi documentation with the above statement, thanks.

**Comment by dlyssenko** [comment]
@gmuloc, I've updated `develop` branch of the documentation: https://pyeapi.readthedocs.io/en/develop/client_modules/client.html#pyeapi.client.connect

**Comment by zohassadar** [comment]
> @gmuloc, I've updated `develop` branch of the documentation: https://pyeapi.readthedocs.io/en/develop/client_modules/client.html#pyeapi.client.connect

When using `return_node=True`, the fix is applied differently:

`cc._connection.transport._context.set_ciphers('DEFAULT')`

**Comment by dlyssenko** [comment]
To close the loop: after the fix #236, a cipher can be set and passed to pyeapi via ssl context - see the same link in the documentation for the update.
